packages/foam-core/src/note-graph.ts

```diff
@@ -1,121 +1,162 @@
-import { Graph, Edge } from 'graphlib';
-import { Position, Point } from 'unist';
-import GithubSlugger from 'github-slugger';
+import { Graph } from 'graphlib';
 import { EventEmitter } from 'events';
+import { Position, Point, URI, ID } from './types';
+import { uriToSlug, hashURI, getUriViaRelative } from './utils';
 
-type ID = string;
-
-export interface Link {
-  from: ID;
-  to: ID;
+export interface NoteSource {
+  uri: URI;
   text: string;
+  contentStart: Point;
+  end: Point;
+  eol: string;
 }
 
-export interface NoteLink {
-  to: ID;
+export interface WikiLink {
+  type: 'wikilink';
   text: string;
+  slug: string;
   position: Position;
 }
 
+export type NoteLink = WikiLink;
+
+export interface NoteInfo {
+  title: string | null;
+  properties: object;
+  // sections: NoteSection[]
+  // tags: NoteTag[]
+  links: NoteLink[];
+  definitions: NoteLinkDefinition[];
+  source: NoteSource;
+}
+
+export type Note = NoteInfo & {
+  id: ID;
+  slug: string; // note: this slug is not necessarily unique
+};
+
+export interface Connection {
+  from: ID;
+  to: ID;
+  link: NoteLink;
+}
+
 export interface NoteLinkDefinition {
   label: string;
   url: string;
   title?: string;
   position?: Position;
 }
 
-export class Note {
-  public id: ID;
-  public frontmatter: object;
-  public title: string | null;
-  public source: string;
-  public path: string;
-  public start: Point;
-  public end: Point;
-  public eol: string;
-  public links: NoteLink[];
-  public definitions: NoteLinkDefinition[];
-
-  constructor(
-    id: ID,
-    frontmatter: object,
-    title: string | null,
-    links: NoteLink[],
-    definitions: NoteLinkDefinition[],
-    start: Point,
-    end: Point,
-    path: string,
-    source: string,
-    eol: string
-  ) {
-    this.id = id;
-    this.frontmatter = frontmatter;
-    this.title = title;
-    this.source = source;
-    this.path = path;
-    this.links = links;
-    this.definitions = definitions;
-    this.start = start;
-    this.end = end;
-    this.eol = eol;
-  }
-}
-
 export type NoteGraphEventHandler = (e: { note: Note }) => void;
 
+export type NoteQuery = {
```

Comment:
hehe you are right. the logic here is that there is a precedence: `id`, `uri`, `slug`, `title`.
atm in case of multiple parameters the above precedence is used to pick the criteria.

I think your idea of having `type` and `value` is interesting, will think about that more.
if I understand correctly it would look like: `getNote(by: <'id' | 'slug' | 'uri' | 'title'>, value: string)`

re more complex queries I would probably go the object approach, but I wonder if that would live in this fn or another (I am ambivalent about this, still need to think about it)